File: sable-shared/src/cache/keys.ts
// All Redis key patterns in one place. Source of truth: gateway DB doc §
// "Redis Keys". Every key the platform reads or writes goes through a
// builder here — no ad-hoc template strings at call sites.
//
// Why functions, not constants: builders catch missing args at compile time
// and prevent typos that silently target the wrong key.

// ---------------------------------------------------------------------------
// Shared entity-type unions (mirror the DB CHECK constraints)
// ---------------------------------------------------------------------------

export type BlockEntityType = 'ip' | 'user_id' | 'org_id' | 'device_fingerprint';
export type WhitelistEntityType = 'admin_account' | 'org' | 'global';
export type RateWindow = 'minute' | 'hour' | 'day';

// ---------------------------------------------------------------------------
// Key builders
// ---------------------------------------------------------------------------

export const cacheKeys = {
  /** Per-user request count. Window keyed separately so we don't reset on hour boundaries. */
  rateUser: (userId: string, window: RateWindow) => `rate:user:${userId}:${window}`,
  /** Per-org aggregate request count (minute window only per the doc). */
  rateOrg: (orgId: string) => `rate:org:${orgId}:minute`,
  /** Per-IP request count (minute window only per the doc). */
  rateIp: (ip: string) => `rate:ip:${ip}:minute`,

  /** Cached active_modules array — invalidated on module change. */
  modulesUser: (userId: string) => `modules:user:${userId}`,

  /** Revoked JWTs — TTL = remaining JWT lifetime. */
  jwtBlacklist: (jti: string) => `blacklist:jwt:${jti}`,

  /** Used HMAC nonces — 30s window matching used_nonces table. */
  nonce: (nonce: string) => `nonce:${nonce}`,

  /** Mirror of active blocked_entities rows. */
  blockCache: (type: BlockEntityType, value: string) => `block:cache:${type}:${value}`,
  /** Mirror of active ip_whitelist rows. */
  whitelistCache: (type: WhitelistEntityType, value: string) => `whitelist:cache:${type}:${value}`,

  /** Cached session state — TTL = remaining session lifetime. */
  session: (sessionId: string) => `session:${sessionId}`,

  /**
   * Hot-path cache: token-hash hex → full session JSON. Lets `authenticate`
   * skip the DB lookup on every request. TTL = remaining session lifetime.
   * Revoke writes invalidate both this and `session(id)` (the service holds
   * the mapping so it can invalidate without the raw token).
   */
  sessionByToken: (tokenHashHex: string) => `session:bytok:${tokenHashHex}`,

  /**
   * API key resolution cache — hash(hex) → owner + scopes. Lets API-key auth
   * skip the DB lookup on warm paths. TTL ≈ 5 min so revokes propagate.
   */
  apiKeyByHash: (keyHashHex: string) => `apikey:bytok:${keyHashHex}`,

  /** Admin session — same pattern, separate cookie namespace. */
  adminSession: (sessionId: string) => `admin-session:${sessionId}`,
  adminSessionByToken: (tokenHashHex: string) => `admin-session:bytok:${tokenHashHex}`,

  /** Bot detection: request frequency counter (sub-50ms inter-request gaps). */
  botRequests50ms: (userId: string) => `bot:requests:${userId}:50ms`,
  /** Bot detection: regularity (auto-driven request patterns). */
  botPatternRegularity: (ip: string) => `bot:pattern:${ip}:regularity`,
  /** Bot detection: interaction telemetry variance (mouse / keyboard). */
  botMouse: (userId: string) => `bot:mouse:${userId}`,

  /** Cached service_routes table for routing. */
  routeCache: () => 'route:cache',
  /** Cached active HMAC key versions. */
  hmacVersions: () => 'hmac:versions',
  /** Cached trusted device fingerprints for a user. */
  fingerprintUser: (userId: string) => `fingerprint:user:${userId}`,
  /** Single key/value cache for `gateway_config` reads. */
  gatewayConfigKey: (key: string) => `config:gw:${key}`,
  /** Cache-bust handle for the whole gateway_config map. */
  configCache: () => 'config:cache',
  /** Cached allowed CORS origins. */
  corsOrigins: () => 'cors:origins',

  // ---------------------------------------------------------------------------
  // Version-stamp keys — `Refreshable<T>` polls these; on diff, the gateway
  // reloads the in-memory map from Postgres. Admin write paths bump the
  // stamp with `Date.now().toString()` to trigger reload across instances.
  // ---------------------------------------------------------------------------
  serviceRoutesVersion: () => 'route:cache:version',
  hmacVersionsVersion: () => 'hmac:versions:version',
  corsOriginsVersion: () => 'cors:origins:version',

  // ---------------------------------------------------------------------------
  // Inbound webhook idempotency — set-once on the external event id so
  // Stripe retries can't double-dispatch the same event.
  // ---------------------------------------------------------------------------
  webhookDedup: (source: string, externalEventId: string) => `webhook:dedup:${source}:${externalEventId}`,

  /** Latest health status snapshot per downstream service. */
  health: (serviceName: string) => `health:${serviceName}`,
} as const;

File: sable-shared-py/sable_shared/cache/keys.py
"""All Redis key patterns in one place. Source of truth: gateway DB doc §
"Redis Keys". Every key the platform reads or writes goes through a
builder here — no ad-hoc f-strings at call sites.

Mirrors TS ``sable-shared/src/cache/keys.ts``. Change one, change both.
"""

from __future__ import annotations

from typing import Literal

# ---------------------------------------------------------------------------
# Shared entity-type unions (mirror the DB CHECK constraints)
# ---------------------------------------------------------------------------

BlockEntityType = Literal["ip", "user_id", "org_id", "device_fingerprint"]
WhitelistEntityType = Literal["admin_account", "org", "global"]
RateWindow = Literal["minute", "hour", "day"]


# ---------------------------------------------------------------------------
# Key builders — module-level functions, all return ``str``.
# ---------------------------------------------------------------------------

def rate_user(user_id: str, window: RateWindow) -> str:
    """Per-user request count. Window keyed separately so we don't reset on
    hour boundaries."""
    return f"rate:user:{user_id}:{window}"


def rate_org(org_id: str) -> str:
    """Per-org aggregate request count (minute window only)."""
    return f"rate:org:{org_id}:minute"


def rate_ip(ip: str) -> str:
    """Per-IP request count (minute window only)."""
    return f"rate:ip:{ip}:minute"


def modules_user(user_id: str) -> str:
    """Cached active_modules array — invalidated on module change."""
    return f"modules:user:{user_id}"


def jwt_blacklist(jti: str) -> str:
    """Revoked JWTs — TTL = remaining JWT lifetime."""
    return f"blacklist:jwt:{jti}"


def nonce(nonce_value: str) -> str:
    """Used HMAC nonces — 30s window matching used_nonces table."""
    return f"nonce:{nonce_value}"


def block_cache(entity_type: BlockEntityType, value: str) -> str:
    """Mirror of active blocked_entities rows."""
    return f"block:cache:{entity_type}:{value}"


def whitelist_cache(entity_type: WhitelistEntityType, value: str) -> str:
    """Mirror of active ip_whitelist rows."""
    return f"whitelist:cache:{entity_type}:{value}"


def session(session_id: str) -> str:
    """Cached session state — TTL = remaining session lifetime."""
    return f"session:{session_id}"


def bot_requests_50ms(user_id: str) -> str:
    """Bot detection: request frequency counter (sub-50ms inter-request gaps)."""
    return f"bot:requests:{user_id}:50ms"


def bot_pattern_regularity(ip: str) -> str:
    """Bot detection: regularity (auto-driven request patterns)."""
    return f"bot:pattern:{ip}:regularity"


def bot_mouse(user_id: str) -> str:
    """Bot detection: interaction telemetry variance (mouse / keyboard)."""
    return f"bot:mouse:{user_id}"


def route_cache() -> str:
    """Cached service_routes table for routing."""
    return "route:cache"


def hmac_versions() -> str:
    """Cached active HMAC key versions."""
    return "hmac:versions"


def fingerprint_user(user_id: str) -> str:
    """Cached trusted device fingerprints for a user."""
    return f"fingerprint:user:{user_id}"


def config_cache() -> str:
    """Cached gateway_config key/value store."""
    return "config:cache"


def cors_origins() -> str:
    """Cached allowed CORS origins."""
    return "cors:origins"


def health(service_name: str) -> str:
    """Latest health status snapshot per downstream service."""
    return f"health:{service_name}"
